### DISCUSSION TITLE
math domain error when plotting a model

### DISCUSSION THREAD

**Discussion by kaisermaneuver** [mjhoptics/ray-optics#119] [Q&A] [answered]
I've been trying to use the library for quite some time and have encountered a couple of issues regarding it:

when trying to add a surface with 
sm.add_surface([1/5.5, 6.0, "N-BK7", "Schott", 1.5168])#camera lens
sm.add_surface([-1/5.5, 55])#camera gap
and then updating it with
opm.update_model()

I get the following error:
 File "/home/audacia/Documentos/vart_lens_2.py", line 69, in <module>
    layout_plt = plt.figure(FigureClass=InteractiveLayout, opt_model=opm, is_dark=isdark).plot()
  File "/home/audacia/.local/lib/python3.10/site-packages/matplotlib/_api/deprecation.py", line 454, in wrapper
    return func(*args, **kwargs)
  File "/home/audacia/.local/lib/python3.10/site-packages/matplotlib/pyplot.py", line 783, in figure
    manager = new_figure_manager(
  File "/home/audacia/.local/lib/python3.10/site-packages/matplotlib/pyplot.py", line 359, in new_figure_manager
    return _get_backend_mod().new_figure_manager(*args, **kwargs)
  File "/home/audacia/.local/lib/python3.10/site-packages/matplotlib/backend_bases.py", line 3504, in new_figure_manager
    fig = fig_cls(*args, **kwargs)
  File "/home/audacia/anaconda3/envs/vessel-env/lib/python3.10/site-packages/rayoptics/mpl/interactivelayout.py", line 57, in __init__
    super().__init__(**kwargs)
  File "/home/audacia/anaconda3/envs/vessel-env/lib/python3.10/site-packages/rayoptics/mpl/interactivefigure.py", line 92, in __init__
    self.update_data()
  File "/home/audacia/anaconda3/envs/vessel-env/lib/python3.10/site-packages/rayoptics/mpl/interactivelayout.py", line 69, in update_data
    self.ele_bbox = self.update_patches(self.ele_shapes)
  File "/home/audacia/anaconda3/envs/vessel-env/lib/python3.10/site-packages/rayoptics/mpl/interactivefigure.py", line 183, in update_patches
    handles = shape.update_shape(self)
  File "/home/audacia/anaconda3/envs/vessel-env/lib/python3.10/site-packages/rayoptics/elem/layout.py", line 86, in update_shape
    self.e.render_handles(self.opt_model)
  File "/home/audacia/anaconda3/envs/vessel-env/lib/python3.10/site-packages/rayoptics/elem/elements.py", line 770, in render_handles
    shape = self.render_shape()
  File "/home/audacia/anaconda3/envs/vessel-env/lib/python3.10/site-packages/rayoptics/elem/elements.py", line 733, in render_shape
    poly1 = full_profile(self.profile1, self.is_flipped, self.extent(),
  File "/home/audacia/anaconda3/envs/vessel-env/lib/python3.10/site-packages/rayoptics/elem/elements.py", line 351, in full_profile
    prf = profile.profile((sd_lwr, sd_upr), dir, steps)
  File "/home/audacia/anaconda3/envs/vessel-env/lib/python3.10/site-packages/rayoptics/elem/profiles.py", line 342, in profile
    adj_start = copysign(sqrt(r*r - sd_start*sd_start), cv)
ValueError: math domain error

Also, I tried to build the model using the elements module, and while I was successful in creating lenses with create_lens and airgaps with create_air_gap, the most I could do is update the element list but not the whole model, much less with the gaps.

Among the examples that I have seen, only the seq_model has been used to create models. Are there any examples of the other modules used to create a model?

**Comment by mjhoptics** [comment, marked as answer]
Hello @kaisermaneuver,
Thank you for trying ray-optics. The trace-back that you attached was generated trying to generate a lens layout. Specifically, the crash occurred trying to evaluate the profile curve of a surface in order to draw it.  This is often caused by something being amiss in the model setup. So I would review the following things, looking for problems.

1) Sequential model (sm). This is the center of most ray-optics models. Use
`sm.list_model()`
to get the basic "radius, thickness, index" definition of the model. If your model has tilts and/or decenters, use:
`sm.list_sg()`
To list everything (aspheres, decenters, phase elements, etc, etc) in the sequential model, use:
`listobj(sm)`
The object distance can be set to **infinity** by setting:
`sm.gaps[0].thi = 1e12`

2) Optical Specification (osp). This can be displayed using:
`listobj(osp)`
Check the pupil and fov specifications for correctness, in particular.
Check the first order properties using the Paraxial Model (pm). To check what the current focal length, magnification, pupil locations and sizes, use:
`pm.first_order_data()`
To list the paraxial marginal and chief ray data, use:
`fo.list_parax_trace()`

3) Do rays trace thru the system? Here you want to trace the axial marginal ray and the chief ray at full field. Suppose your model has 3 fields. This will trace the marginal and chief rays:
```
axial_fld, wvl, foc = osp.lookup_fld_wvl_focus(0)
list_ray(trace_base(opm, [0, 1], axial_fld, wvl))
edge_fld, wvl, foc = osp.lookup_fld_wvl_focus(2)
list_ray(trace_base(opm, [0, 0], edge_fld, wvl))
```

These are the initial steps that I would follow when trying to debug a crashing model. If you have any questions about what the output means, please ask! This helps me also when adding to the documentation, to fill in holes. Feel free to post the output for help debugging the model.
 Hope this helps,
Mike Hayford

**Reply by kaisermaneuver** [reply]
Hi Mike

Thank you for the response. I checked the pupil and fov specifications for correctness, in particular, and got the following results

aperture: object pupil; value=10.0
field: object angle; value=0.0
x,y=0.0 vlx= 0.000 vux= 0.000 vly= 0.000 vuy= 0.000
central wavelength=587.5618 nm
wavelength (weight) =  486.1327 (0.500),  587.5618 (1.000)*,  656.2725 (0.500)
focus shift=0.0

However, the value of field: object angle should be 120°, given that I defined:
osp.pupil = PupilSpec(osp, key=['object', 'pupil'], value=10.0) 
osp.field_of_view = FieldSpec(osp, key=['object', 'angle'], value=120,  flds=[0.],is_relative=False)
osp.spectral_region = WvlSpec([(486.1327, 0.5), (587.5618, 1.0), (656.2725, 0.5)], ref_wl=1)

Really appreciate any insight on this
Steffen

**Comment by mjhoptics** [comment]
Hello Steffen,
Thanks for the response. I see one issue and that is the field angle. In `ray-optics`, as in other codes, the half-angle is used to specify the field value. So a spec of 120 deg means a total field angle of 240 degrees, a hyper-hemispherical field of view. I haven't tested `ray-optics` for fields beyond 90 degrees and it does not surprise me that they don't work "right out of the box".  If you really meant a total field of 120 degrees, the field value should be entered as 60.

Also, the field specification is often a list of fields, e.g. axis, 0.7 field and full field. The entry would look like this:
```
osp.field_of_view = FieldSpec(osp, key=['object', 'angle'], value=60, flds=[0., 42., 60.],is_relative=False)
```
Note that the full field value is entered as both the field value and in the list of fields. Alternatively, using relative specification:
```
osp.field_of_view = FieldSpec(osp, key=['object', 'angle'], value=60, flds=[0., .707, 1.], is_relative=True)
```
Hope this helps; feel free to ask more questions!
MIke

**Reply by kaisermaneuver** [reply]
Hi Mike

Thanks for the response. I adjusted accordingly and got some progress, even listed the objects and got the information I was looking for. However, I'm still having issues trying to plot the system. Below is the script I'm working on, where the uncommented add_surface gets plotted with and the ray tracing is strange to say the least; when I try to replicate with the commented system, the layout never comes and still get the math domain error

```
import matplotlib.pyplot as plt
isdark = False

import itertools
from rayoptics.util.misc_math import euler2opt
import transforms3d as t3d

from rayoptics.environment import *

from rayoptics.elem.elements import Element, Mirror
from rayoptics.elem import parttree

from rayoptics.util.misc_math import normalize

opm = OpticalModel()
sm  = opm['seq_model']
osp = opm['optical_spec']
pm = opm['parax_model']
em = opm['ele_model']
pt = opm['part_tree']
ar = opm['analysis_results']

osp.pupil = PupilSpec(osp, key=['object', 'pupil'], value=10.0) 
osp.field_of_view = FieldSpec(osp, key=['object', 'angle'], value=60,  flds=[0., 40., 60.],is_relative=False)
#osp.field_of_view = FieldSpec(osp, key=['object', 'angle'], value=60,  flds=[0., .707, 1.],is_relative=False)
osp.spectral_region = WvlSpec([(486.1327, 0.5), (587.5618, 1.0), (656.2725, 0.5)], ref_wl=1)
opm.system_spec.dimensions = 'MM'
opm.radius_mode = True

sm.gaps[0].thi=1e10

#sm.add_surface([0., 0.])
#sm.set_stop()

#####working system
sm.add_surface([-2.182, .75, 1.603, 27.5])
sm.add_surface([-6.367, 0.5])
sm.add_surface([10.50, 5.0, 1.500, 62.5])
sm.add_surface([-10.50, .65517241])
sm.add_surface([-5.00, 2.0, 1.500, 62.5])
sm.add_surface([5.00, .65517241])


#sm.add_surface([1/5.5, 6.0, "N-BK7", "Schott"])#camera lens
#sm.add_surface([-1/5.5, 55])#camera gap
#sm.add_surface([1/10.5, 3.0, "N-BK7", "Schott"])#moving lens
#sm.add_surface([-1/10.5, 7.5])#moving gap
#sm.add_surface([1/10, 3.0, "N-BK7", "Schott"])#vqped lens
#sm.add_surface([-1/10, 36.8])#vqped gap


opm.update_model()
sm.list_model()
listobj(sm)
listobj(osp)

live_layout = plt.figure(FigureClass=InteractiveLayout, opt_model=opm,do_draw_rays=True, do_paraxial_layout=False,offset_factor=0.1, is_dark=isdark).plot()
```

Really appreciate any insight on this
Steffen

**Reply by mjhoptics** [reply]
Hi Steffen,

Thanks for the more detailed example. The issue I see here is the pupil specification. The PupilSpec is defining a 10mm entrance pupil diameter whereas the radius of the first surface is 2.18. So the pupil diameter is defined to be twice the possible diameter of the first surface. This is the immediate cause of the ray failures.

If I change the pupil diameter to be 1mm, I get the following lens layout:

<img width="660" alt="Screen Shot 2023-05-17 at 12 58 14 PM" src="https://github.com/mjhoptics/ray-optics/assets/33294031/be261eae-0798-42a3-a913-6579197cbe0d">

Further Notes
--------------
The second, commented out, specification, can be coerced into tracing by setting radius_mode to False -> meaning add_surface() inputs are interpreted as curvatures.  Here's the code:
```
opm = OpticalModel()
sm  = opm['seq_model']
osp = opm['optical_spec']
pm = opm['parax_model']
em = opm['ele_model']
pt = opm['part_tree']
ar = opm['analysis_results']

#osp.pupil = PupilSpec(osp, key=['object', 'pupil'], value=10.0) 
osp.pupil = PupilSpec(osp, key=['object', 'pupil'], value=1.0) 
#osp.field_of_view = FieldSpec(osp, key=['object', 'angle'], value=60,  flds=[0., 40., 60.],is_relative=False)
osp.field_of_view = FieldSpec(osp, key=['object', 'angle'], value=30,  flds=[0., .707, 1.],is_relative=True)
osp.spectral_region = WvlSpec([(486.1327, 0.5), (587.5618, 1.0), (656.2725, 0.5)], ref_wl=1)
opm.system_spec.dimensions = 'MM'

opm.radius_mode = False  #  <--- add_surface data seems to be curvatures

sm.gaps[0].thi=1e10

#sm.add_surface([0., 0.])
#sm.set_stop()

#####working system
#sm.add_surface([-2.182, .75, 1.603, 27.5])
#sm.add_surface([-6.367, 0.5])
#sm.add_surface([10.50, 5.0, 1.500, 62.5])
#sm.add_surface([-10.50, .65517241])
#sm.add_surface([-5.00, 2.0, 1.500, 62.5])
#sm.add_surface([5.00, .65517241])

sm.add_surface([1/5.5, 6.0, "N-BK7", "Schott"])#camera lens
sm.add_surface([-1/5.5, 55])#camera gap
sm.add_surface([1/10.5, 3.0, "N-BK7", "Schott"])#moving lens
sm.add_surface([-1/10.5, 7.5])#moving gap
sm.add_surface([1/10, 3.0, "N-BK7", "Schott"])#vqped lens
sm.add_surface([-1/10, 36.8])#vqped gap

opm.update_model()
```
The lens layout looks like this:
![issue119](https://github.com/mjhoptics/ray-optics/assets/33294031/045dc87a-776c-4a94-aa4d-03b21f29d155)

Hope this helps.

Mike
